testnet/stacks-node/src/nakamoto_node/miner.rs

```diff
@@ -0,0 +1,621 @@
+                coinbase_tx,
+                // TODO (refactor): the nakamoto block builder doesn't use this VRF proof,
+                //  it has to be included in the coinbase tx, which is an arg to the builder.
+                //  we should probably just remove this from the nakamoto block builder.
+                vrf_proof: vrf_proof.clone(),
+                tenure_change_tx,
+            })
+        } else {
+            None
+        };
+
+        parent_block_info.stacks_parent_header.microblock_tail = None;
+
+        // build the block itself
+        let (mut block, _, _) = match NakamotoBlockBuilder::build_nakamoto_block(
+            &chain_state,
+            &burn_db.index_conn(),
+            &mut mem_pool,
+            // TODO (refactor): the nakamoto block builder doesn't use the parent tenure ID,
+            //  it has to be included in the tenure change tx, which is an arg to the builder.
+            //  we should probably just remove this from the nakamoto block builder, so that
+            //  there isn't duplicated or unused logic here
+            &self.parent_tenure_id,
+            &parent_block_info.stacks_parent_header,
+            &self.burn_block.consensus_hash,
+            self.burn_block.total_burn,
+            tenure_start_info,
+            self.config.make_block_builder_settings(
+                // TODO: the attempt counter needs a different configuration approach in nakamoto
+                1,
+                false,
+                self.globals.get_miner_status(),
+            ),
+            Some(&self.event_dispatcher),
+        ) {
+            Ok(block) => block,
+            Err(e) => {
+                error!("Relayer: Failure mining anchored block: {}", e);
+                return None;
+            }
+        };
+
+        let mining_key = self.keychain.get_nakamoto_sk();
+        let miner_signature = mining_key
+            .sign(block.header.signature_hash().ok()?.as_bytes())
+            .ok()?;
+        block.header.miner_signature = miner_signature;
+
+        info!(
+            "Miner: Succeeded assembling {} block #{}: {}, with {} txs",
+            if parent_block_info.parent_block_total_burn == 0 {
+                "Genesis"
+            } else {
+                "Stacks"
+            },
+            block.header.chain_length,
+            block.header.block_hash(),
+            block.txs.len(),
+        );
+
+        // last chance -- confirm that the stacks tip is unchanged (since it could have taken long
+        // enough to build this block that another block could have arrived), and confirm that all
+        // Stacks blocks with heights higher than the canoincal tip are processed.
+        let cur_burn_chain_tip = SortitionDB::get_canonical_burn_chain_tip(burn_db.conn())
+            .expect("FATAL: failed to query sortition DB for canonical burn chain tip");
+
+        if cur_burn_chain_tip.consensus_hash != block.header.consensus_hash {
```

Comment:
You wouldn't want to compare to the block's consensus hash, because the sortition may have been empty and the current miner's tenure may still be ongoing.  Would it make more sense to compare the sortition tips before and after building the block?

Reply:
I'm not sure yet how the nakamoto MinerThread should handle the sortition changing. My current plan is to treat it like a tenure change and do a thread stop and start. That way, the `MinerThread` really should stop mining if it detects a consensus hash change (which it will sort of need to do in any event: it will have to get a new tenure change tx).